Feature: SwiftUI Notification System

boredarthur opened the pull request.

boredarthur committed:
Introduced Notification System

boredarthur committed:
toasts, toasts queues

boredarthur committed:
refactoring, tests

boredarthur committed:
Merge branch 'chore/swift-6-1-migration' into feature/swiftui-notification-system

boredarthur committed:
confirmation dialog

boredarthur committed:
Merge branch 'chore/swift-6-1-migration' into feature/swiftui-notification-system

boredarthur committed:
renamed notification to dialog

boredarthur committed:
revert alert builder

boredarthur committed:
deprecations

Maks-Jago committed:
Merge branch '1.5.0' into feature/swiftui-notification-system

boredarthur committed:
renamed `DialogState` to `DialogStatus`

boredarthur committed:
alert backwards compitability

Maks-Jago commented on UDF/View/Dialog/Themes/ToastTheme.swift:
```diff
@@ -0,0 +1,375 @@
+    ///   - infoColors: New info color styling.
+    /// - Returns: A new theme with updated colors.
+    func withColors(
+        error errorColors: ColorStyle? = nil,
```
нащо множина? errorColor без s?

boredarthur replied:
Множина, тому що `ColorStyle` є множиною різних категорій кольорів. Можна задати як один певний колір `.solid(.red)`, так і декілька кольорів у градієнті `.gradient([.red, .pink, .orange])`

Maks-Jago commented on UDF/View/Dialog/Themes/ToastTheme.swift:
```diff
@@ -0,0 +1,375 @@
+    /// Defines the background color treatment for error states. Can be solid colors
+    /// or gradients. Should convey urgency and attention while remaining accessible.
+    /// - Default: Solid red (.red)
+    public var errorColors: ColorStyle
```
тут та нижче чому в множині?

boredarthur replied:
З такої ж причини як описав вище

Maks-Jago commented on UDF/View/Dialog/Themes/Shared/ToastPosition.swift:
```diff
@@ -0,0 +1,248 @@
+        case .center:
+            hasher.combine("center")
+        case .custom:
+            hasher.combine("custom")
```
краще передавати усі параметри в hasher у випадку custom.

boredarthur replied:
`Alignment` не є Hashable типом, а `CGPoint` вважається Hashable лише з iOS 18

Maks-Jago replied:
Alignment в стрінгу, стрінгу в хешер, те саме і з CGPoint

boredarthur replied:
Done

Maks-Jago commented on UDF/View/Dialog/Themes/Shared/ToastIcon.swift:
```diff
@@ -0,0 +1,274 @@
+    /// - Keep view complexity minimal for smooth animation performance
+    /// - Avoid heavy computations or network calls in view content
+    /// - Consider caching complex views when used frequently
+    case view(AnyView)
```
Пропоную прибрати case view з ToastIcon взагалі, а в DialogContent зробити доп ініт в якому буде кастомна в'юха для іконки. В підсумку буде 

```swift
DialogContent<Icon: View>:
...
let iconView: @Sendable () -> Icon
```

в тих інітах де іконка пишеш 
init where IconView == Image


а ToastIcon виглядає так як взагалі можна видалити

Maks-Jago commented on UDF/View/Dialog/Themes/Shared/ColorStyle.swift:
```diff
@@ -0,0 +1,227 @@
+                    startPoint: .leading,
+                    endPoint: .trailing
+                )
+            }
```
може так, щоб більш чистіше було
```swift
case .gradient(let colors) where colors.isEmpty: Color.clear
case .gradient(let colors) where colors.count == 1: colors[0]
default:
   LinearGradient(
                    colors: colors,
                    startPoint: .leading,
                    endPoint: .trailing
                )
```

Maks-Jago commented on UDF/View/Dialog/Modifiers/ToastDialogModifier.swift:
```diff
@@ -0,0 +1,70 @@
+        case .dismissed:
+            // Intentionally left empty - individual toasts manage their own dismissal
+            // We don't want to clear all toasts when a single dialog is dismissed
+            break
```
а хіба ми не повинні зачищати/видаляти чергу тостів якщо нам сказали "сховайтесь"?

Maks-Jago commented on UDF/View/Dialog/Modifiers/ToastDialogModifier.swift:
```diff
@@ -0,0 +1,70 @@
+            }
+            .onAppear {
+                updateToastPresentation(dialogStatus)
+            }
```
А якщо appear тригірнеться N разів, ми N разів покажемо один і той самий тост?

Maks-Jago commented on UDF/View/Dialog/Modifiers/ToastDialogModifier.swift:
```diff
@@ -0,0 +1,70 @@
+                updateToastPresentation(dialogStatus)
+            }
+            .overlay {
+                ToastContainer() { _ in }
```
чому тут { _ in } ?

Maks-Jago commented on UDF/View/Dialog/Modifiers/ConfirmationDialogModifier.swift:
```diff
@@ -0,0 +1,110 @@
+    let cancelButton: DialogButton?
+    let destructiveButtons: [DialogButton]
+    let regularButtons: [DialogButton]
+}
```
а нащо нам зберігати кнопки по ролям окремо?

Maks-Jago commented on UDF/View/Dialog/Modifiers/ConfirmationDialogModifier.swift:
```diff
@@ -0,0 +1,110 @@
+            }
+            .onAppear {
+                updateDialogStatus()
+            }
```
а так як раніше формувати на льоту в body не підходить? чи безпечніше на appear та мати опціонали?

<img width="1076" alt="Screenshot 2025-06-19 at 16 54 03" src="https://github.com/user-attachments/assets/9794b147-c9f3-4ec7-be74-0030d83cf7db" />

Maks-Jago commented on UDF/View/Dialog/Modifiers/ConfirmationDialogModifier.swift:
```diff
@@ -0,0 +1,110 @@
+                                cancelButton.action()
+                                dismissDialog()
+                            }
+                        }
```
Давай використовувати уніфікований підхід з AlertAction протоколом як то було з Алертом. Змінити назву на DialogAction а принцип такий самий. Де потім конкретний action конформить протокол і задає своє body щоб потім використати тут так само як було в алертах
<img width="559" alt="Screenshot 2025-06-19 at 17 05 52" src="https://github.com/user-attachments/assets/15d71b60-d593-4483-b5a2-69dd74c23f8b" />

Maks-Jago commented on UDF/View/Dialog/Modifiers/AlertDialogModifier.swift:
```diff
@@ -0,0 +1,126 @@
+                ),
+                actions: {
+                    if let alertState = alertState {
+                        ForEach(Array(alertState.actions.enumerated()), id: \.offset) { _, action in
```
Давай використовувати старий підхід як було з Алертами де в нас DialogAction сама формує своє body а тут ми потім явно використовуємо кнопку і все

boredarthur replied:
Це також було виправлено

Maks-Jago commented on UDF/View/Dialog/Modifiers/AlertDialogModifier.swift:
```diff
@@ -0,0 +1,126 @@
+/// alert system and preserves all existing alert behavior and presentation logic.
+struct AlertDialogModifier: ViewModifier {
+    @Binding var dialogStatus: DialogStatus
+    @State private var alertState: AlertState?
```
в такій реалізації якщо алерт показаний на екрані і хтось тригерне переманлювання, то алерт блимне (сховається, покажеться). В минулій реалізації це фіксилось за допомогою localAlert'у та кастомного isPresented.
В такому вигляді такої проблеми з алертом що блимати буде немає?

boredarthur replied:
Це було поправлено

Maks-Jago commented on UDF/View/Dialog/Core/DialogType.swift:
````diff
@@ -0,0 +1,189 @@
+/// ```
+public enum DialogType: Hashable, Sendable {
+    /// A success dialog with a message.
+    case success(String, style: DialogStyle)
````
давай явно додамо message всюди.
```swift
case success(**message**: String, style: DialogStyle)
```

Maks-Jago commented on UDF/View/Dialog/Content/DialogContent.swift:
```diff
@@ -0,0 +1,460 @@
+    /// - Ensure custom views maintain proper accessibility labels and hints
+    /// - Test with VoiceOver and other assistive technologies
+    /// - Provide appropriate contrast ratios for text and interactive elements
+    public nonisolated(unsafe) let customView: AnyView?
```
також прибрати AnyView, зробивши так само як описав нижче з ToastIcon

boredarthur committed:
colors refactoring

boredarthur committed:
smart toasts dismiss

boredarthur committed:
prevent duplicate toast enqueuing

boredarthur committed:
handle toast dismissal callback properly

boredarthur committed:
rename message parameter in DialogType cases

boredarthur committed:
removed buttons role separation

boredarthur committed:
modifier render on the fly

boredarthur committed:
refactoring of actions

boredarthur committed:
removed toast icon

boredarthur committed:
refactored view generic

boredarthur committed:
Merge branch '1.5.0' into feature/swiftui-notification-system

Maks-Jago committed:
refactor dialog generic

Maks-Jago committed:
update toast view/container

Maks-Jago committed:
remove commented code

boredarthur committed:
refactoring

boredarthur committed:
hasher

Maks-Jago committed:
Merge pull request #72 from Maks-Jago/chore/dialog-generic-refactoring

Dialog generic refactoring

Maks-Jago commented on Tests/SwiftUI-UDF-Tests/Dialog/DialogTests.swift:
```diff
@@ -0,0 +1,390 @@
+            
+            // Check custom icon
+            XCTAssertTrue(content.hasIcon)
+//            XCTAssertEqual(content.icon, .systemImage("party.popper.fill"))
```
Треба по тестам пройтись і підправити

boredarthur replied:
Done

Maks-Jago commented on Tests/SwiftUI-UDF-Tests/Dialog/DialogTests.swift:
```diff
@@ -0,0 +1,390 @@
+        
+        // Verify it's a toast with custom view
+        if case .presented(let dialogType) = status {
+//           case .custom(let content, let style) = dialogType  {
```
тут також

boredarthur replied:
Done

Maks-Jago commented on UDF/Common/Extensions/View+AlertStatus.swift:
```diff
@@ -15,216 +15,17 @@ import SwiftUI
     ///
+    /// **Deprecated**: This method now internally delegates to the new Dialog system.
+    /// Please migrate to using `DialogStatus` with `.dialog(status:)` for new code.
+    /// This method will be removed in version 1.5.1.
```
давай писати "in future updates"

boredarthur replied:
Done

Maks-Jago commented on UDF/Store/Action/Actions.swift:
```diff
@@ -67,16 +68,17 @@ public enum Actions {
+    
     /// `UpdateAlertStatus` is an action used to update the status of an alert within the UDF architecture.
     /// It contains the alert's status and an identifier, enabling the management of alerts based on their unique IDs.
+    @available(*, deprecated, message: "Will be removed in version 1.5.1. Use UpdateDialogStatus instead.")
```
також давай писати про future updates без конкретних версій

boredarthur replied:
Done

Maks-Jago committed:
refactor alert actions to use dialog actions and clean up

Removed AlertAction, AlertButton, AlertTextField, and AlertActionsBuilder in favor of DialogAction, DialogButton, DialogTextField, and DialogActionsBuilder. Updated AlertBuilder and related code to use the new dialog-based types and builders. Added deprecation typealiases for backward compatibility. Updated function signatures and type constraints to use 'any' and '@Sendable' where appropriate, and improved consistency across dialog and alert APIs.

Maks-Jago committed:
Update DialogAction.swift

Maks-Jago commented on UDF/View/Dialog/Modifiers/AlertDialogModifier.swift:
```diff
@@ -0,0 +1,160 @@
+                                Button(button.title, role: button.role) {
+                                    button.action()
+                                    handleAlertDismissal()
+                                }
```
Тут кнопки явно створювати не можна, ми повинні рендерити body DialogButton'а як це було в попередній реалізації

boredarthur replied:
Done

Maks-Jago committed:
Update ConfirmationDialogModifier.swift

Maks-Jago commented on UDF/View/Dialog/Modifiers/AlertDialogModifier.swift:
```diff
@@ -0,0 +1,160 @@
+                            if let button = action as? DialogButton {
+                                Button(button.title, role: button.role) {
+                                    button.action()
+                                    handleAlertDismissal()
```
нащо нам тут handleAlertDismissal? треба обходитись без доступу до action кнопки. додатково глянь як було

boredarthur replied:
Прибрано

Maks-Jago committed:
Merge pull request #73 from Maks-Jago/chore/alert-to-dialog

Chore/alert to dialog

boredarthur committed:
updated deprecation messages

boredarthur committed:
render button using body

boredarthur committed:
removed body call

boredarthur committed:
fixed tests

Maks-Jago commented on UDF/View/Dialog/Modifiers/AlertDialogModifier.swift:
```diff
@@ -40,11 +40,7 @@ struct AlertDialogModifier: ViewModifier {
-                                    handleAlertDismissal()
-                                }
-                                .disabled(button.disabled)
+                                button.body
```
просто button, нащо body?

boredarthur replied:
<img width="620" height="182" alt="CleanShot 2025-07-16 at 15 42 16" src="https://github.com/user-attachments/assets/eaa1742a-9e9e-44e2-9bda-22a8b77cada3" />
Це щось гітхаб шалить, тут немає body

Maks-Jago commented on UDF/View/Dialog/Actions/DialogAction.swift:
```diff
@@ -0,0 +1,117 @@
+
+/// Protocol for classifying dialog action types and capabilities.
+/// This is used internally for presentation logic.
+internal protocol DialogActionClassification {
```
internal писати не треба, він йде за замовчуванням

boredarthur replied:
Прибрав

Maks-Jago commented on UDF/View/Dialog/Configurations/ToastQueueConfiguration.swift:
```diff
@@ -0,0 +1,557 @@
+        
+        // Process queue for next toast
+        if configuration.displayMode == .sequential {
+            Task {
```
а на якому потоці працює ToastQueueConfiguration? чи не потрібно тут та вище додати @MainActor ?

boredarthur replied:
Так, потрібно. Додав

Maks-Jago commented on UDF/View/Dialog/Configurations/ToastQueueConfiguration.swift:
```diff
@@ -0,0 +1,557 @@
+
+        guard duration > 0 else { return }
+        
+        let task = Task {
```
і тут

boredarthur replied:
Done

boredarthur committed:
minor changes
